addons/mail/models/mail_template.py

```diff
@@ -342,12 +344,13 @@ def _generate_template_recipients(self, res_ids, render_fields,
 
                 if not all_emails:
                     continue
-                additional_values = {}
-                if company_id:
-                    additional_values['company_id'] = company_id
                 partners = self.env['res.partner']._find_or_create_from_emails(
                     all_emails,
-                    additional_values=additional_values
+                    additional_values=lambda norm_email: {
+                        **ModelSudoPrefetch.browse(email_to_res_ids.get(norm_email, [])
+                                                   )._get_customer_information().get(norm_email, {}),
```

Comment:
Still have this issue that in a 4K record loop, we are going to call 4K times the same method. We could call it once and access it when necessary instead ... actually, is there any gain of giving a callable versus giving a directly a dict with {'email_normalized': values} directly ? That way _find_or_create_from_emails seems simpler to me. But maybe you have good reasons to prefer a callable ?

Reply:
Moreover if we build a final dict, we can remove the company-based loop I think (loop here above) and call directly everything in batch, unless I misunderstood something.

The tweak to do would be to support a default value when there is no email normalized (what is done here using the callable, but with a "False" key in the dict for example).

As you prefer, but I find the callable a bit over engineered.